### ISSUE TITLE
Clone in word embedding

### ISSUE THREAD

**Issue by luo3300612** [ruotianluo/self-critical.pytorch#197]
Thank you for your repo. I have a question about `clone` before word embedding. Here is code in AttModel.py
```python
for i in range(seq.size(1)):
    if self.training and i >= 1 and self.ss_prob > 0.0: # otherwiste no need to sample
        sample_prob = fc_feats.new(batch_size*seq_per_img).uniform_(0, 1)
        sample_mask = sample_prob < self.ss_prob
        if sample_mask.sum() == 0:
            it = seq[:, i].clone()
        else:
            sample_ind = sample_mask.nonzero().view(-1)
            it = seq[:, i].data.clone()
            prob_prev = torch.exp(outputs[:, i-1].detach()) # fetch prev distribution: shape Nx(M+1)
            it.index_copy_(0, sample_ind, torch.multinomial(prob_prev, 1).view(-1).index_select(0, sample_ind))
```
I have read document and learn that `tensor.clone()` returns a copy of original tensor and gradient of cloning tensor will be add to original tensor.
Since seq is ground truth caption, its `require_grad` property is False. So 
1.why we need to clone seq[:, i] when sample_mask.sum()==0? Is it to avoid changing it unintendedly later?
2.What's the difference between `.clone()` and `.data.clone()` in schedule sampling?

**Comment by ruotianluo** [comment]
I use clone because it's the safest choice. Same with data.clone
